python/openassetio_traitgen/generators/cpp.py

```diff
@@ -0,0 +1,297 @@
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+"""
+A traitgen generator that outputs a C++ package based on the
+openassetio_traitgen PackageDefinition model.
+"""
+
+import logging
+import os
+import re
+
+from typing import List
+
+import jinja2
+
+from . import helpers, cpp_keywords
+from ..datamodel import PackageDeclaration, PropertyType
+
+
+__all__ = ["generate"]
+
+
+#
+## Code Generation
+#
+
+
+# pylint: disable=too-many-locals
+def generate(
+    package_declaration: PackageDeclaration,
+    globals_: dict,
+    output_directory: str,
+    creation_callback,
+    logger: logging.Logger,
+):
+    """
+    Generates a C++ package for the supplied definition under
+    output_directory.
+    """
+
+    env = _create_jinja_env(globals_, logger)
+
+    def render_template(name: str, path: str, variables: dict):
+        """
+        A convenience to render a named template into its corresponding
+        file and call the creation_callback.
+        """
+        # pylint: disable=line-too-long
+        # NB: Jinja assumes '/' on all plaftorms:
+        #  https://github.com/pallets/jinja/blob/7fb13bf94443f067c74204a1aee368fdf0591764/src/jinja2/loaders.py#L29
+        template = env.get_template(f"cpp/{name}.hpp.in")
+        with open(path, "w", encoding="utf-8") as file:
+            file.write(template.render(variables))
+        creation_callback(path)
+
+    def create_dir_with_path_components(*args) -> str:
+        """
+        A convenience to create a directory from the supplied path
+        components, calling the creation_callback and returning its path
+        as a string.
+        """
+        path = os.path.join(*args)
+        os.makedirs(path, exist_ok=True)
+        creation_callback(path)
+        return path
+
+    # Top level package directory, under an "include" subdirectory
+    package_name = env.filters["to_cpp_module_name"](package_declaration.id)
+    package_dir_path = create_dir_with_path_components(
+        output_directory, package_name, "include", package_name)
+
+    # Collect which sub-packages we should import at the top level, so
+    # they're available without needing to `#include` each individual
+    # header file.
+    package_init_imports = []
+
+    # Sub-packages for traits and specifications
+    for kind in ("traits", "specifications"):
+
+        namespaces = getattr(package_declaration, kind, None)
+        if namespaces:
+
+            package_init_imports.append(f"{kind}/{kind}.hpp")
+
+            # Create the directory for the sub-package
+            subpackage_dir_path = create_dir_with_path_components(package_dir_path, kind)
+
+            # Collect the resulting module names for each namespace
+            # So we can pre-import them in the sub-package init.
+            subpackage_init_imports = []
+
+            # Generate a single-file module for each namespace
```

Comment:
> We should at least be a bit cleverer with cross-package trait references in Specifications, though, if possible, for the reason you highlight. Currently the implementation does similar to Python, in that it imports the entire external package.

Specification headers also `#include` the entire `traits.hpp` of their package, which again could be unwieldy if you just want a specific MediaCreation specification (or rather, specification namespace) that only depends on a subset of traits.

Refactoring this will be a bit of work, since we get into dependency-tree territory, which I think is worth a separate PR/issue.

Reply:
Capturing off-line discussions:
 - Splitting to headers is valuable to allow those with a strict 'include what you use' philosophy to use it.
 - Hoists are fine
 - Longer term may wish to move away from hoists in Specifications.

Reply:
Added a file-per-class in dd3deba1 then refactored the `generate` function in d966e630